### ISSUE TITLE
Discussion About How To Handle Null Values

### ISSUE THREAD

**Issue by AndydeCleyre** [wwkimball/yamlpath#98]
## Operating System

1. Name/Distribution: Arch

## Version of Python and packages in use at the time of the issue.

1. [Distribution](https://wiki.python.org/moin/PythonDistributions): CPython
2. Python Version: 3.8.6
3. Version of yamlpath installed: 3.3.0
4. Version of ruamel.yaml installed: 0.16.12

## Minimum sample of YAML (or compatible) data necessary to trigger the issue

Smallest:
```yml
nullthing: null
```

A little bigger:
```yml
---
nullthing: null
intthing: 6
floatthing: 6.8
yesthing: yes
nothing: no
truething: true
falsething: false
```

JSON:
```json
{
  "falsething": false,
  "floatthing": 6.8,
  "intthing": 6,
  "nothing": "no",
  "nullthing": null,
  "truething": true,
  "yesthing": "yes"
}
```

## Complete steps to reproduce the issue when triggered via:

1. Command-Line Tools (yaml-get, yaml-set, or eyaml-rotate-keys):

```console
% yaml-get -p nullthing nully.yml  # or:
% yaml-get -p .nullthing nully.yml  # or:
% yaml-get -p '[. == nullthing]' nully.yml  # etc.
```
2. Libraries (yamlpath.*):  I've not yet used yamlpath as a library, just CLI.

## Expected Outcome

I don't know! Some thoughts:

- Usually single results are printed "bare" (without any type indicators like quotes), so I don't know how to distinguish between an empty string and null result, or between "null" and `null`.
- An empty string may be alright.
- Maybe a non-zero return code is indeed appropriate :shrug: .
- This may overlap with a separate wish: option to treat everything other than a hash or list as a string (no native numbers, booleans, dates, or nulls).

## Actual Outcome

```
CRITICAL:  Required YAML Path does not match any nodes
```
(retcode 1)

## Screenshot(s), if Available

![image](https://user-images.githubusercontent.com/1787385/98145634-048b4e00-1e99-11eb-99ef-6229aac8c284.png)

**Comment by wwkimball** [comment]
Null (AKA None, AKA nil) is the absence of value.  It is not equivalent to an empty-string, which is actually *something*; an empty-string is a character-string which happens to have zero length and thus, no characters.  A null is similarly not zero.  Nothing matches null values because there's nothing there to compare against; there are even some programming languages which won't even allow null to be compared against null.  Reporting an empty-string as a result would be technically incorrect.

This is tricky.  The YAML Specification doesn't define `key: null` as a key named `key` with a `null` value.  Rather, it stipulates that `key: null` be interpreted as a key named `key` with a bare string value of, `"null"`.  It is interesting that Python (and thus, yamlpath) has elected to translate a bare string of `"null"` into a `None` value (Python's version of a `null`).  More importantly, the YAML specification clearly stipulates that a key with a null value must be represented as `key:` (nothing after the `:`).  See example "2.21 Miscellaneous" at https://yaml.org/spec/1.2/spec.html#id2761292 for the official specification on this.

At present, the code reports when it retrieves *anything*, pass/fail.  Since null is the opposite of *anything*, the code reports that there was nothing to pull from the document.  I appreciate that users may expect *something* when they deliberately query a known null node, but the code receives *nothing* from the query -- not even an answer of a single empty node -- and interprets that as there was never a node there to query.  I could change the code to be sensitive to receiving perhaps nothing but an indication that the target node's parent exists, but that takes us back to:  how do we represent a null result in the output?

This is difficult to call a "bug" because there isn't a viable expected outcome.  It could be an enhancement if we could agree on how to handle this case.

That said, there is a way to use YAML-spec-correct null values:  YAML Tags.  The next version of yamlpath has support for YAML Tags along with some other goodies like creating, assigning, and renaming YAML Anchors and Aliases from the command-line (I keep a [CHANGES log in the development branch](https://github.com/wwkimball/yamlpath/blob/development/CHANGES) which always shows what I'm working on).  I bring this up because we can use a YAML Tag for null values:  `!null`.  When I tried this just now, I got this result:

**Document**:
``yaml
---
nullthing: !null
```

**Query via bash**:
```text
$ echo -e "---\nnullthing: "'!'"null" | yaml-get --query=nullthing; echo $?

0
```

**Query via PowerShell**:
```text
echo "---`nnullthing: !null`n" | yaml-get --query=/nullthing; echo $?

True
```

As you can see, a blank line was reported for the null result; not an empty-string, just a line with nothing printed on it.  The command also reported 0 rather than 1.

I'm open to ideas.  What do you think?

**Comment by AndydeCleyre** [comment]
That all sounds reasonable, and I don't think I can offer any good advice here.

I won't personally use tags, so maybe my own answer is along the lines of just not querying for `null`s, and/or using strictyaml as a preprocessor for yamlpath (and working out a json->yaml preprocessor for strictyaml...).

---

For reference I'm including here another strictyaml comparison, not using any schema:

```python
#!/usr/bin/env python3
import json
import sys
from pprint import pprint

import strictyaml as sy
from plumbum.cmd import yaml_get


if __name__ == '__main__':
    if sys.argv[1:]:
        with open(sys.argv[1]) as yfile:
            original = yfile.read()
    else:
        original = sys.stdin.read()
    yaml = sy.load(original)

    print("---original:", original, sep='\n')

    print("---strictyaml interprets this as dict:")
    pprint(dict(yaml.data))

    print("\n---strictyaml re-renders this as yaml:")
    print(yaml.as_yaml())

    print("---strictyaml->json:")
    print(json.dumps(yaml.data))

    jcontent = (yaml_get['-p', '.'] <<original)()
    print("\n---yaml-get interprets original as:", jcontent, sep='\n')
```

`---original:`
```yml
---
nullthing: null
intthing: 6
floatthing: 6.8
yesthing: yes
nothing: no
truething: true
falsething: false
nothingthing:
```
`---strictyaml interprets this as dict:`
```python
{'falsething': 'false',
 'floatthing': '6.8',
 'intthing': '6',
 'nothing': 'no',
 'nothingthing': '',
 'nullthing': 'null',
 'truething': 'true',
 'yesthing': 'yes'}
```
`---strictyaml re-renders this as yaml:`
```yml
nullthing: null
intthing: 6
floatthing: 6.8
yesthing: yes
nothing: no
truething: true
falsething: false
nothingthing:
```
`---strictyaml->json:`
```json
{"nullthing": "null", "intthing": "6", "floatthing": "6.8", "yesthing": "yes", "nothing": "no", "truething": "true", "falsething": "false", "nothingthing": ""}
```
`---yaml-get interprets original as:`
```json
{"nullthing": null, "intthing": 6, "floatthing": 6.8, "yesthing": "yes", "nothing": "no", "truething": true, "falsething": false, "nothingthing": null}
```

**Comment by wwkimball** [comment]
I'm happy to explore options.  I'm a little confused with your example code above.  It looks to me like you got every bit of data back out of the sample file, including the nulls.  I don't see any difference between the yaml-get output any any others.  In fact, the yaml-get data looks better than strictyaml->json because the data-types are preserved (which is correct for JSON).

If we work backwards from your end-goal, we might find the solution you need.  So, what exactly are you trying to do with the data that you presently cannot?  It has to be more than just "get a null value", because you can do that right now as long as you select the node's parent, which your example above demonstrates.

**Comment by wwkimball** [comment]
I experimented with allowing yaml-get to just return a blank line for a result that is nothing and a 0 exit state instead of the error message that no nodes matched.

This looks like:

```text
$ yaml-get --query=nullthing andy-deCleyre-nulls.yaml; echo $?

0
```

However, this is wrong because it is indistinguishable from an empty-string node, like:

```yaml
---
nullthing: null
intthing: 6
floatthing: 6.8
yesthing: yes
nothing: no
truething: true
falsething: false
nothingthing:
emptystring: ''
```

```text
$ yaml-get --query=emptystring andy-deCleyre-nulls.yaml; echo $?

0
```

I similarly cannot return `null` as a result in this case because that is indistinguishable from a node for which the value is a string of `"null"`.

This is tricky, indeed...

**Comment by wwkimball** [comment]
I just had an idea:  You could always just take the result as JSON, for which null values are represented as a non-string `null` literal.  **Did you know you can use Collectors to convert any results to a list, which forces the output to be in JSON form?**

Try this:
```text
yaml-get --query='(nullthing)' andy-deCleyre-nulls.yaml; echo $?
[null]
0
```

Does that help?

**Comment by wwkimball** [comment]
Here's an idea I might run with:  Just use the standard ASCII/ANSI NULL special control character (which is a non-printing hexadecimal \x00).  This has the advantage of being a well-established (and very old) standard.  It creates a parsible result; you can read the NULL character down-stream from your query.  It also gives us something to return so the above "Collect the result as JSON" trick works to get the JSON literal, `null`, when preferred.

To a human, the result will look like a blank line and a 0 exit-state.  However, that "blank line" will actually have a NULL control-character on it before the CR/LF control-character(s), which non-human parsers will pick up.

Unless you object, I'm going to move forward with this.

**Comment by AndydeCleyre** [comment]
Sorry I haven't been able to offer more at this time. The null character does seem like a good match for representing a null value in a shell or script.

**Comment by wwkimball** [comment]
I have implemented this change into the next release.  There's still some work to be done on some other features but this improvement will appear in version 3.4.0.

Thank you for working with me on this!!

**Comment by AndydeCleyre** [comment]
Will the next release's existing changes allow for accessing empty-string results in general yaml-get queries? I'm happy to open a new discussion issue about it, but will not generate the noise if these related changes already cover the case.

To be clearer:

`vars.yml`:
```yml
spacings:
  - ''  # wide glyphs more often
  - term  # mono more often
  - force-monospace  # term with no wide glyphs
  - fixed  # force-monospace with no ligatures
```

I'm looping over the `spacings` [in a shell script](https://github.com/AndydeCleyre/archbuilder_iosevka/blob/cf570bab57927670f69cc57f3154fd037359ad69/mk/pkgbuilds.sh), with:

```sh
yaml-get -p 'spacings[. =~ //]' "${gitroot}"/vars.yml | while read spacing; do
```

The above works, but the only reason I'm using a regex match rather than, say, `spacings.*`, is that the latter does not currently include the empty string value in the output.

**Comment by wwkimball** [comment]
There are two answers to your question:

1. This behavior is [detailed in the Wiki](https://github.com/wwkimball/yamlpath/wiki/Wildcard-Segments#match-every-immediate-child) and [mentioned in the README](https://github.com/wwkimball/yamlpath/blob/master/README.md#supported-yaml-path-segments).  The single-star wildcard is short-hand for a search, making it easier to write search segments than using the `[...]` syntax.  The default search expression -- when there is only a single star with no adjoining phrases -- is `[.!=""]`.  That translates specifically to, "return every immediate child whose value is *not empty-string*".  The Wiki goes on to offer that if you want empty-string nodes (along with every nested child), use the double-star segment, `**`, instead.  Unlike the single-star segment type, the double-star is not a search; rather, it deeply traverses the DOM from that point and yields every leaf node, regardless its value.

1. Just because that's how I wrote it on the last pass doesn't mean that's the absolute best way to handle single-star segments.  I set the default to `[.!=""]` only because I didn't consider anyone would ever **want** empty-string results.  And yet, here you are, wanting exactly that.  :)

I will change the default single-star search expression from `[.!=""]` to `[.=~/.*/]` (there's nothing wrong with using a RegEx; this is a search short-hand, anyway), which should yield every single immediate child (no deep traversal), no matter what its value is.  I'll update the README and Wiki accordingly.

Thanks for keeping me on my toes!  :D

**Comment by AndydeCleyre** [comment]
Ah, thanks so much! Yes I need to read the docs better. This is a great tool, in no small part due to your active consideration.

**Comment by wwkimball** [comment]
I have released version 3.4.0, which includes all of the above changes along with some other new capabilities.  I hope this helps with your projects!

**Comment by wwkimball** [comment]
Afterthought (sorry, this really just came to me when I wasn't even thinking about any of this):  I suddenly remembered "working out a json->yaml preprocessor" from one of your earlier posts.  I don't know if you're aware, but yamlpath can convert documents between YAML, JSON, and back.  The yaml-merge command has a `--document-format` (`-D`) option which enables such conversions.  Since yaml-merge can operate against a single document -- nothing to merge, just a doc to read -- you can use it to handle conversations.

For example:

```shell
$ cat some_file.yaml 
---
this: file
started: as YAML
but:
  was: converted
  to: JSON
and:
  - back
  - again

$ yaml-merge --document-format=json --output=some_file.json some_file.yaml 
Processing some_file.yaml...

$ cat some_file.json 
{"this": "file", "started": "as YAML", "but": {"was": "converted", "to": "JSON"}, "and": ["back", "again"]}

$ yaml-merge --document-format=yaml some_file.json
---
"this": "file"
"started": "as YAML"
"but":
  "was": "converted"
  "to": "JSON"
"and":
  - "back"
  - "again"
```

The present version preserves all demarcation marks in the JSON document, so the resulting YAML-from-JSON looks a little ugly to the eye but it is valid.

Hope this helps in some way!
